app/README.md

```diff
@@ -0,0 +1,221 @@
+| `DBG_SRV` | `0` | DBG-level for server, passed as `-debug` to the `conode` command |
+| `NBR` | `3` | Highest number of servers and clients |
+| `NBR_SERVERS` | `NBR` | How many servers to configure |
+| `NBR_SERVERS_GROUP` | `NBR_SERVERS - 1` | How many servers to write to the
+group-file - per default, keep one server out of the group, to test what
+happens if a server gets added later |
+| `APPDIR` | `pwd` | where the `test.sh`-script is located |
+| `APP` | `$(basename $APPDIR)` | The name of the builddir |
+
+All variables can be overriden by defining them _before_ the inclusion of
+`libtest.sh`. The most common setup of a `test.sh` only has the first two
+global variables, `DBG_TEST` and `DBG_SRV`:
+
+```bash
+# Show the output of the commands: 0=none, 1=test-names, 2=all
+DBG_TEST=1
+# DBG-level for call to the app - this is not handled by libtest.sh, but will
+# be used in the helper methods.
+DBG_APP=2
+# DBG-level for server
+DBG_SRV=0
+
+. $(go env GOPATH)/src/github.com/dedis/onet/app/libtest.sh
+```
+
+If a test fails, it is common to change those variables to:
+
+```bash
+DBG_TEST=2
+DBG_SRV=2
+```
+
+So that `libtest.sh` shows the output of the CLI command and the `conode` prints
+at a debug-level of 2.
+
+### `main`
+
+To have the most important method at the top (for easier editing), `main`
+is defined at the beginning. So any edits to tests can be done there.
+
+A typical main looks like this:
+
+```bash
+main(){
+    startTest
+    buildConode
+    test Build
+    test Network
+    stopTest
+}
+```
+
+* `startTest` cleans the test-directory and builds the CLI binary
+* `buildConode` creates a new conode and automatically includes the
+service defined in the `./service` or `../service` directory
+* `test Build` makes sure all conodes are stopped and removes all databases
+from the previous test, then calls `testBuild`
+* `test Network` makes sure all conodes are stopped and removes all databases
+from the previous test, then calls `testNetwork`
+* `stopTest` stops all remaining conodes and cleans up the test-directory, if it
+is the temporary directory
+
+One common use-case of the `main` method in the case of a failing test is to
+comment out all tests that run successfully up to the failing test, so that
+a subsequent `./test.sh` run only runs the failing test. Calling `./test.sh -nt`
+allows for changing the failing `testName` method to find out what is wrong.
+
+### Test-methods
+
+Each test-method usually starts the required conodes before running the tests
+necessary to verify the correct running of the service.
+
+```bash
+testNetwork(){
+    runCoBG 1 2
+    testOut "Running network"
+    testGrep "Available_Services" runCl -g public.toml
+    testGrep "Available_Services" runCl -g public.toml
+}
+```
+
+* `runCoBG 1 2` starts conode number `1` and `2` (the index starts at 1) in
+the background. There is some code to make sure the conodes are actually up
+and running and listening on the websocket-ports.
```

Comment:
Is there a reason why we don't have `runCoBG 2` to mean start 2 conodes, similar to `run_conode.sh`? Is it because we define the number of conodes in `$NBR`? This function confused me in the past.